### ISSUE TITLE
RuntimeError: "addmm_impl_cpu_" not implemented for 'Half

### ISSUE THREAD

**Issue by azuryl** [huggingface/peft#1597]
### System Info

peft v0.10

### Who can help?

@pacman100 @younesbelkada @sayakpaul

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```
Start Evaluating
mv: cannot stat 'tune_log/llama_3B_0.2_param_second_block_wise_fusion0.6_N50_Dora_r8/checkpoint-1400/pytorch_model.bin': No such file or directory
Selected Tasks: ['boolq', 'winogrande', 'arc_easy', 'piqa', 'arc_challenge', 'openbookqa', 'hellaswag']
Load from Pruned Model!!!!!!!!!!: prune_log/llama_3B_prune_0.2_block_wise_param_second_fusion0.6_N50/pytorch_model.bin
##lm_eval/huggingface.py _create_auto_model_peft dtype!!!! None
lm_eval/huggingface.py self.AUTO_PEFT_CLASS.from_pretrained torch_dtype torch.float16
LLMpruner#####layer.py Linear init use_dora : True
Traceback (most recent call last):
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/main.py", line 116, in <module>
    main()
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/main.py", line 81, in main
    results = evaluator.simple_evaluate(
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/lm_eval/utils.py", line 182, in _wrapper
    return fn(*args, **kwargs)
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/lm_eval/evaluator.py", line 64, in simple_evaluate
    lm = lm_eval.models.get_model(model).create_from_arg_string(
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/lm_eval/base.py", line 115, in create_from_arg_string
    return cls(**args, **args2)
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/lm_eval/models/huggingface.py", line 232, in __init__
    self.model = self._create_auto_model_peft(
  File "/data/azuryl/LLM-Pruner/lm-evaluation-harness/lm_eval/models/huggingface.py", line 296, in _create_auto_model_peft
    model = self.AUTO_PEFT_CLASS.from_pretrained(
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/peft_model.py", line 355, in from_pretrained
    model = MODEL_TYPE_TO_PEFT_MODEL_MAPPING[config.task_type](model, config, adapter_name)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/peft_model.py", line 1094, in __init__
    super().__init__(model, peft_config, adapter_name)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/peft_model.py", line 129, in __init__
    self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/model.py", line 136, in __init__
    super().__init__(model, config, adapter_name)
  File "/data/azuryl/LLM-Pruner/peft/tuners/tuners_utils.py", line 148, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "/data/azuryl/LLM-Pruner/peft/tuners/tuners_utils.py", line 325, in inject_adapter
    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/model.py", line 220, in _create_and_replace
    new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/model.py", line 295, in _create_new_module
    new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/layer.py", line 1057, in dispatch_default
    new_module = Linear(target, adapter_name, **kwargs)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/layer.py", line 357, in __init__
    self.update_layer(
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/layer.py", line 126, in update_layer
    self.dora_init(adapter_name)
  File "/data/azuryl/LLM-Pruner/LLMPruner/peft/tuners/lora/layer.py", line 191, in dora_init
    lora_weight = lora_B.weight @ lora_A.weight  #.float()  
RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
```

### Expected behavior

how to fix 
if I am not changeto   lora_weight = lora_B.weight.float() @ lora_A.weight.float()

**Comment by BenjaminBossan** [comment]
Could you please give us more context? Ideally, you could paste the code you're using that leads to this error. What system are you using, CPU or GPU? Please also share the versions of the other libraries you use (torch, transformers, etc.).

**Comment by azuryl** [comment]
ubuntu 20.04 use GPU
transformers 4.29.1
torch version 2.1.2+cu121


export PYTHONPATH='.'
python lm-evaluation-harness/main.py --model hf-causal-experimental \
       --model_args checkpoint=PATH_TO_PRUNE_MODEL,peft=PATH_TO_SAVE_TUNE_MODEL,config_pretrained=PATH_OR_NAME_TO_BASE_MODEL \
       --tasks openbookqa,arc_easy,winogrande,hellaswag,arc_challenge,piqa,boolq \
       --device cuda:0 --no_cache \
       --output_path PATH_TO_SAVE_EVALUATION_LOG

**Comment by azuryl** [comment]
main.py


import argparse
import json
import logging
import fnmatch
import time
import sys
from lm_eval import tasks, evaluator

logging.getLogger("openai").setLevel(logging.WARNING)


class MultiChoice:
    def __init__(self, choices):
        self.choices = choices

    # Simple wildcard support (linux filename patterns)
    def __contains__(self, values):
        for value in values.split(","):
            if len(fnmatch.filter(self.choices, value)) == 0:
                return False

        return True

    def __iter__(self):
        for choice in self.choices:
            yield choice


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", required=True)
    parser.add_argument("--model_args", default="")
    parser.add_argument("--tasks", default=None, choices=MultiChoice(tasks.ALL_TASKS))
    parser.add_argument("--provide_description", action="store_true")
    parser.add_argument("--num_fewshot", type=int, default=0)
    parser.add_argument("--batch_size", type=str, default=None)
    parser.add_argument("--device", type=str, default=None)
    parser.add_argument("--output_path", default=None)
    parser.add_argument("--limit", type=int, default=None)
    parser.add_argument("--no_cache", action="store_true")
    parser.add_argument("--decontamination_ngrams_path", default=None)
    parser.add_argument("--description_dict_path", default=None)
    parser.add_argument("--check_integrity", action="store_true")

    return parser.parse_args()


# Returns a list containing all values of the source_list that
# match at least one of the patterns
def pattern_match(patterns, source_list):
    task_names = set()
    for pattern in patterns:
        for matching in fnmatch.filter(source_list, pattern):
            task_names.add(matching)
    return list(task_names)


def main():
    args = parse_args()

    assert not args.provide_description  # not implemented

    if args.limit:
        print(
            "WARNING: --limit SHOULD ONLY BE USED FOR TESTING. REAL METRICS SHOULD NOT BE COMPUTED USING LIMIT."
        )

    if args.tasks is None:
        task_names = tasks.ALL_TASKS
    else:
        task_names = pattern_match(args.tasks.split(","), tasks.ALL_TASKS)

    print(f"Selected Tasks: {task_names}")

    description_dict = {}
    if args.description_dict_path:
        with open(args.description_dict_path, "r") as f:
            description_dict = json.load(f)

    start = time.time()
    results = evaluator.simple_evaluate(
        model=args.model,
        model_args=args.model_args,
        tasks=task_names,
        num_fewshot=args.num_fewshot,
        batch_size=args.batch_size,
        device=args.device,
        no_cache=args.no_cache,
        limit=args.limit,
        description_dict=description_dict,
        decontamination_ngrams_path=args.decontamination_ngrams_path,
        check_integrity=args.check_integrity,
    )
    end = time.time() - start
    print("@@@@@@@@@@@Latency:",end)
    dumped = json.dumps(results, indent=2)
    print(dumped)

    if args.output_path:
        import os
        directory_path = os.path.dirname(args.output_path)
        if not os.path.exists(directory_path):
            os.makedirs(directory_path)

        with open(args.output_path, "w") as f:
            f.write(dumped)

    print(
        f"{args.model} ({args.model_args}), limit: {args.limit}, provide_description: {args.provide_description}, "
        f"num_fewshot: {args.num_fewshot}, batch_size: {args.batch_size}"
    )
    print(evaluator.make_table(results))


if __name__ == "__main__":
    main()

**Comment by BenjaminBossan** [comment]
Thanks for providing further information. Since your script depends on custom checkpoints, I'm still not able to reproduce it. From what I can tell, the model is initialized on CPU first (probably the device is only applied later by lm-evaluation-harness) and the dtype for the model is half (float16), which causes this error. I tried to isolate this specifically, but on my machine, initializing a DoRA model on CPU with float16 works.

For your situation, could you use a different dtype, like bfloat16 or float32 + AMP? That may solve the issue you have.

**Comment by aliftadvantage** [comment]
I'm running into this issue when training SDXL using --use-dora in the dreambooth training script. And using load_lora_weights() to load the weights.

**Comment by BenjaminBossan** [comment]
> I'm running into this issue when training SDXL using --use-dora in the dreambooth training script. And using load_lora_weights() to load the weights.

This seems to be a little bit of a different use case than the original one. For your problem, do you have a CUDA-capable GPU? Does the same script work when you don't use DoRA? What dtype are you using -- try different values of `--mixed_precision` and `--prior_generation_precision`.

**Comment by BenjaminBossan** [comment]
> I use fp16, it since fp16 just for cuda.

Be aware though that the error message you pasted indicates that this code is running on CPU, which has not support for fp16.

**Comment by azuryl** [comment]
> > I use fp16, it since fp16 just for cuda.
> 
> Be aware though that the error message you pasted indicates that this code is running on CPU, which has not support for fp16.

yes, at begin code run on GPU

**Comment by BenjaminBossan** [comment]
Okay, I was investigating further as I was unable to reproduce your error. It appears that the operation is not supported on CPU by older PyTorch versions. Can you try to upgrade to PyTorch v2.2 and see if that fixes your issue? Otherwise, is it possible to run the code on GPU right from the start?

It would be possible for PEFT to provide a workaround for your situation in case those solutions don't work for you, but please try them first.

**Comment by BenjaminBossan** [comment]
> It would be possible for PEFT to provide a workaround for your situation in case those solutions don't work for you, but please try them first.

Okay, we merged #1653, this should fix the issue.
